### DISCUSSION TITLE
Different spinner increment via pressing arrow-up and page-up

### DISCUSSION THREAD

**Discussion by Croissant-Tsai** [ControlSystemStudio/phoebus#3159] [Q&A]
Hello,

I am new to Phoebus, and I am trying to figure out how to make a spinner increment by different values when pressing the arrow-up key (+0.1) and the page-up key (+1). I have tried to find a possible function and written a script to listen for key-pressing in org.csstudio.display.builder, but it doesn't seem to be working. Could anyone help me? Thanks!

**Comment by kasemir** [comment]
The spinner (and slider, and scrollbar) have only one `Increment` property. They do not differentiate between cursor up/down or page up/down. Besides, Mac keyboards don't even have dedicated page keys.

Maybe try the thumbwheel widget, there you can separately control the ones, tens, hundreds etc.

**Reply by Croissant-Tsai** [reply]
Thank you for your response!

Unfortunately, the thumbwheel does not meet our requirements. We need to control over 100 electrodes in the microscope, so having both large and small step adjustments, as well as a number-entry function, is important for us to tune the image quality.

Additionally, I noticed that in the older version of CS-Studio (4.5.9), the spinner widget has a property called "Page Increment," which allows for two different step sizes.

Could you also let me know if there is a possible solution to listen for keyboard and mouse inputs so that I can control the increment step in the script. Thank you once again :)

![scrot_071](https://github.com/user-attachments/assets/a38c6718-583f-46ce-81b7-35427989b5f0)

**Reply by tynanford** [reply]
At our facility, the users would not accept phoebus if it meant they lost their USB knob controllers. So we created this screen:

![knob-screen](https://github.com/user-attachments/assets/6cc6da30-19ce-482b-870d-f43d79d4c9ad)

Listening for keyboard/mouse inputs in phoebus didn't seem easy/feasible so we created this somewhat hacky but workable solution:

- created real PVs for each item/action on the display. "LeftGain", "LeftKeyUp", "LeftKeyDown", "LeftPVName", etc.
- connected the USB knobs to a mac mini without a monitor that runs a python script that listens for key events, and then writes to these knob panel PVs with pyepics
- Added a bunch of rules to the phoebus screen to listen for changes and update the widget properties. So if the user hits "GAIN UP", the "LeftGain" PV value changes from 1e-3 to 1e-2 and the `Increment` property on the slider widget is changed via a phoebus rule
- Include buttons on other screens to assign specific devices to the left or right knob. This updates the "LeftPVName" PV which in turn updates the slider PVName property via a rule

**Comment by georgweiss** [comment]
I guess the "knob events" are not seen at all in Java/Phoebus, right? If so then a small native component might solve the issue. What kind of knob controller hardware are we talking about? Links to documentation?

But maybe this is not an issue.

**Reply by tynanford** [reply]
The knob controllers we have include software which allows you to map knob events to keyboard events. So turn the knob left can be mapped to pressing page down. Or it could be mapped to pressing the "t" key or whatever. Then in python it's easy to listen for keypress events

**Reply by georgweiss** [reply]
Ok, I see.
